### PR TITLE
Intensive care uat age range

### PR DESCRIPTION
**Pull request by metas-rc**
https://github.com/metasfresh/metasfresh/issues/12570

### CODE REVIEW

File: backend/de.metas.handlingunits.base/src/main/java/de/metas/handlingunits/age/AgeAttributesService.java [hunk 1]
```diff
@@ -94,4 +84,67 @@ public int computeDefaultAge()
 
 		return attributesRepo.retrieveAttributeValues(age);
 	}
+
+	public List<Object> getSuitableValues(final Set<BPartnerId> bPartnerIds, final Set<ProductId> productIds, final Object attributeValue)
```
Comment (metas-ts) [inline backend/de.metas.handlingunits.base/src/main/java/de/metas/handlingunits/age/AgeAttributesService.java:88]: LE: maybe reaname the method toExtractMatchingValues? might be misleading..i'm not sure what the method does..but "get" makes it sound more trivial than it is..

File: backend/de.metas.handlingunits.base/src/main/java/de/metas/storage/spi/hu/impl/HUStorageQuery.java [hunk 2]
```diff
@@ -271,6 +275,17 @@ public IStorageQuery addAttribute(
 		final List<Object> attributeValues = Collections.singletonList(attributeValue);
 		huQueryBuilder.addOnlyWithAttributeInList(attribute, attributeValueType, attributeValues);
 
+		if (HUAttributeConstants.ATTR_Age.equals(AttributeCode.ofString(attribute.getValue())))
+		{
+			final List<Object> ageValues = ageAttributesService.getSuitableValues(getBPartnerIds(), getProductIds(), attributeValue);
```
Comment (metas-ts) [inline backend/de.metas.handlingunits.base/src/main/java/de/metas/storage/spi/hu/impl/HUStorageQuery.java:280]:
```suggestion
			// "explode" the age-attribute value to a range, using before- and after-intervals from the masterdata
			final List<Object> ageValues = ageAttributesService.getSuitableValues(getBPartnerIds(), getProductIds(), attributeValue);
```

**Review by metas-ts — CHANGES_REQUESTED** [review]
..generally, please annotable new parameters (or existing ones that you find along the way) with `@NonNull` or `@Nullable` to reduce the fear of NPEs

**Comment by metas-ts** [comment]
Later edit:
please also provide a field description to make the fields better understandable. I think to understand what it's about, the fieldname is not enough. Please also take into account the infos you gave to kay for him to come up with the field name, and also the infos from the issue's requirements.

**Comment by metas-ts** [comment]
Yes later edit:
I couldn'T find the code for for the "picking profile (v2) config to consider attributes"..
i'm not really sure what it does..i suspect that when i know, maybe i will suggest to drop it, or ask for a clearer field name and maybe also a field-sescription, to reduce FUD

File: backend/de.metas.handlingunits.base/src/main/java/de/metas/handlingunits/age/AgeAttributesService.java [hunk 3]
```diff
@@ -94,4 +84,67 @@ public int computeDefaultAge()
...
+	{
+		int pickingAgeTolerance_BeforeMonths = 0;
+		int pickingAgeTolerance_AfterMonths = 0;
+
+		for (final ProductId productId : productIds) // #12570 there should be only one product here. Not sure how to decide which are the time ranges if there are more
```
[the remark this answers is not in the document]
Comment (metas-rc) [reply backend/de.metas.handlingunits.base/src/main/java/de/metas/handlingunits/age/AgeAttributesService.java:93]: done

File: backend/de.metas.handlingunits.base/src/main/java/de/metas/handlingunits/age/AgeAttributesService.java [hunk 4]
```diff
@@ -94,4 +84,67 @@ public int computeDefaultAge()
...
+		{
+			final I_M_Product product = productDAO.getById(productId);
+
+			final int productBefore = product.getPicking_AgeTolerance_BeforeMonths();
```
[the remark this answers is not in the document]
Comment (metas-rc) [reply backend/de.metas.handlingunits.base/src/main/java/de/metas/handlingunits/age/AgeAttributesService.java:97]: refactored

File: backend/de.metas.handlingunits.base/src/main/java/de/metas/handlingunits/age/AgeAttributesService.java [hunk 5]
```diff
@@ -94,4 +84,67 @@ public int computeDefaultAge()
...
+					pickingAgeTolerance_BeforeMonths = bpartnerProductBefore;
+
+					final int bpartnerProductAfter = bPartnerProduct.getPicking_AgeTolerance_AfterMonths();
+
+					pickingAgeTolerance_AfterMonths = bpartnerProductAfter;
+				}
+			}
+		}
+
+		final int minimumValue = Integer.parseInt(attributeValue.toString()) - pickingAgeTolerance_BeforeMonths;
```
[the remark this answers is not in the document]
Comment (metas-rc) [reply backend/de.metas.handlingunits.base/src/main/java/de/metas/handlingunits/age/AgeAttributesService.java:130]: done

File: backend/de.metas.handlingunits.base/src/main/java/de/metas/handlingunits/age/AgeAttributesService.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (metas-rc) [reply backend/de.metas.handlingunits.base/src/main/java/de/metas/handlingunits/age/AgeAttributesService.java:88]: done

**Review by metas-ts — APPROVED** [review]

**Pull request merged**